### PR TITLE
PR: AADEntitlementManagementAccessPackageAssignmentPolicy - Initial Release

### PR DESCRIPTION
**Pull request by William-Francillette**
#### This Pull Request (PR) fixes the following issues

AADEntitlementManagementAccessPackageAssignmentPolicy : Initial release

I'm finally able to post this PR - I had quite a few challenges with the cmdlets and had to manage the case of the parameters 
It should be all working now

### CODE REVIEW

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 1]
```diff
@@ -0,0 +1,1684 @@
...
+            }
+            $formattedCustomExtensionHandlers += $customExt
+        }
+        #endregion
+
+        #region Format ExpirationDateTime
+        $formattedExpirationDateTime=$getValue.ExpirationDateTime
+        if($null -ne $getValue.ExpirationDateTime)
+        {
+            #$formattedExpirationDateTime=$getValue.ExpirationDateTime.ToString()
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:288]: Why is this If-statement here, if the statement doesn't contain any code.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 2]
```diff
@@ -0,0 +1,1684 @@
...
+            ManagedIdentity         = $ManagedIdentity.IsPresent
+        }
+
+        return [System.Collections.Hashtable] $results
+    }
+    catch
+    {
+        try
+        {
+            Write-Verbose -Message $_
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:322]: Please update this logging logic with the same code as is currently being used in other resources. This has recently been updated to create better logging.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 3]
```diff
@@ -0,0 +1,1684 @@
...
+        #region resource generator code
+        #$UpdateParameters.add('accessReviewSettings',$null)
+        #$UpdateParameters.add('expirationDateTime',$null)
+
+        write-verbose ($UpdateParameters|convertto-json -depth 20)
+
+        Set-MgEntitlementManagementAccessPackageAssignmentPolicy `
+            -BodyParameter $UpdateParameters `
+            -AccessPackageAssignmentPolicyId $currentInstance.Id
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:524]: Please remove extra empty lines

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 4]
```diff
@@ -0,0 +1,1684 @@
...
+        #endregion
+
+    }
+    elseif ($Ensure -eq 'Absent' -and $currentInstance.Ensure -eq 'Present')
+    {
+        Write-Verbose -Message "Removing the access package assignment policy {$DisplayName}"
+        #region resource generator code
+        Remove-MgEntitlementManagementAccessPackageAssignmentPolicy -AccessPackageAssignmentPolicyId $currentInstance.Id
+        #endregion
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:535]: Please remove extra empty line

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 5]
```diff
@@ -0,0 +1,1684 @@
...
+        $RequestApprovalSettings,
+
+        [Parameter()]
+        [Microsoft.Management.Infrastructure.CimInstance]
+        $RequestorSettings,
+
+        [Parameter()]
+        [Microsoft.Management.Infrastructure.CimInstance[]]
+        $CustomExtensionHandlers,
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:55]: Please remove extra empty lines

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 6]
```diff
@@ -0,0 +1,1684 @@
...
+        }
+    }
+
+    $ValuesToCheck.Remove('Credential') | Out-Null
+    $ValuesToCheck.Remove('ApplicationId') | Out-Null
+    $ValuesToCheck.Remove('TenantId') | Out-Null
+    $ValuesToCheck.Remove('ApplicationSecret') | Out-Null
+    $ValuesToCheck.Remove('Id') | Out-Null
+
+    #Write-Verbose -Message "Current Values: $(Convert-M365DscHashtableToString -Hashtable $CurrentValues)"
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:680]: Why are these lines commented out? They are in all resource for extra troubleshooting info.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 7]
```diff
@@ -0,0 +1,1684 @@
...
+
+    try
+    {
+
+        #region resource generator code
+        [array]$getValue = Get-MgEntitlementManagementAccessPackageAssignmentPolicy `
+            -All `
+            -ErrorAction Stop
+        #endregion
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:752]: Please remove extra empty line

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 8]
```diff
@@ -0,0 +1,1684 @@
...
+            Write-Host $Global:M365DSCEmojiGreenCheckMark
+        }
+        return $dscContent
+    }
+    catch
+    {
+        Write-Host $Global:M365DSCEmojiRedX
+        try
+        {
+            write-Host  ($_|out-string)
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:966]: Please update this logging with the new logging code that was changed recently

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 9]
```diff
@@ -0,0 +1,1684 @@
...
+        [System.String]
+        $Key,
+
+        [Parameter(Mandatory = 'true')]
+        $Value,
+
+        [Parameter()]
+        [System.String]
+        $Space="                "
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:1397]: Please remove empty line

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 10]
```diff
@@ -0,0 +1,1684 @@
...
+                        -ReferenceObject ($referenceObject) `
+                        -DifferenceObject ($differenceObject)
+                }
+
+                if ($null -ne $compareResult)
+                {
+                    Write-Verbose -Message "Configuration drift - simple object key: $key Source {$sourceValue} Target {$targetValue}"
+                    return $false
+                }
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:1626]: Please remove empty lines

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/readme.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+
+# AADEntitlementManagementAccessPackageAssignmentPolicy
+
+## Description
+
+This resource configures an A A D Entitlement Management Access Package Assignment Policy.
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/readme.md:6]: Please rename A A D to AzureAD

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/settings.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
+{
+    "resourceName": "AADEntitlementManagementAccessPackageAssignmentPolicy",
+    "description": "This resource configures an A A D Entitlement Management Access Package Assignment Policy.",
+    "permissions":    {
+    "graph":  {
+                  "delegated":  {
+                                    "read":  [
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/settings.json:8]: Please add the required permissions:
EntitlementManagement.Read.All

https://learn.microsoft.com/en-us/graph/api/accesspackageassignmentpolicy-get?view=graph-rest-1.0&tabs=http

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/settings.json [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+                                             ],
+                                    "update":  [
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/settings.json:11]: Please add the required permission:
EntitlementManagement.ReadWrite.All

https://learn.microsoft.com/en-us/graph/api/entitlementmanagement-post-assignmentpolicies?view=graph-rest-1.0&tabs=http

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/settings.json [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+                                               ]
+                                },
+                  "application":  {
+                                      "read":  [
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/settings.json:16]: Please add the required  permissions:
EntitlementManagement.Read.All

https://learn.microsoft.com/en-us/graph/api/accesspackageassignmentpolicy-get?view=graph-rest-1.0&tabs=http

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/settings.json [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+                                               ],
+                                      "update":  [
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/settings.json:19]: Please add the required permission:
EntitlementManagement.ReadWrite.All

https://learn.microsoft.com/en-us/graph/api/entitlementmanagement-post-assignmentpolicies?view=graph-rest-1.0&tabs=http

**Review by ykuijs — CHANGES_REQUESTED** [review]
There are a lot of lines commented out in the code. If these aren't used anymore, why are they in there? Can't they be removed?

Also, please add an example for the resource. That makes using this resource easier for users.

And don't forget to update the Changelog

**Comment by William-Francillette** [comment]
It should be all done now - I was too eager to post this one and forgot to remove my debugging traces
I've added 2 examples as well

**Comment by ykuijs** [comment]
No worries, will review it later today. Just one comment though: Still missing an update in the changelog 😉

**Comment by William-Francillette** [comment]
Ooops, that one slipped through 

Changelog updated 😁

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 16]
```diff
@@ -0,0 +1,1572 @@
...
+    {
+        $ConnectionMode = New-M365DSCConnection -Workload 'MicrosoftGraph' `
+            -InboundParameters $PSBoundParameters `
+            -ProfileName 'beta'
+
+        Select-MgProfile 'beta'
+    }
+    catch
+    {
+        Write-Verbose -Message ($_)
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:95]: Why a Verbose message here? If creating a connection fails, it doesn't make sense to run the rest of the code.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 17]
```diff
@@ -0,0 +1,1572 @@
...
+    {
+        $ConnectionMode = New-M365DSCConnection -Workload 'MicrosoftGraph' `
+            -InboundParameters $PSBoundParameters `
+            -ProfileName 'beta'
+
+        Select-MgProfile 'beta' -ErrorAction Stop
+    }
+    catch
+    {
+        Write-Verbose -Message $_
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:408]: Same comment as in the Get method

**Review by ykuijs — CHANGES_REQUESTED** [review]
Just a small comment

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 (hunk 16 above)
Comment (William-Francillette) [reply Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1]: Sorted

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 (hunk 17 above)
Comment (William-Francillette) [reply Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1]: Sorted

**Comment by NikCharlebois** [comment]
@William-Francillette can you look at resolving the changelog conflicts. Looks like it's pulling from many locations. Thanks

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 18]
```diff
@@ -0,0 +1,1559 @@
...
+        $RequestorSettings,
+
+        [Parameter()]
+        [Microsoft.Management.Infrastructure.CimInstance[]]
+        $CustomExtensionHandlers,
+
+        [Parameter(Mandatory = $true)]
+        [System.String]
+        [ValidateSet('Absent', 'Present')]
+        $Ensure = $true,
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:58]: Default value is incorrect. Should be either Absent or Present, not True

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1 [hunk 19]
```diff
@@ -0,0 +1,1559 @@
...
+
+        [Parameter()]
+        [Microsoft.Management.Infrastructure.CimInstance]
+        $RequestorSettings,
+
+        [Parameter()]
+        [Microsoft.Management.Infrastructure.CimInstance[]]
+        $CustomExtensionHandlers,
+
+        [Parameter(Mandatory = $true)]
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.psm1:55]: Like other PR, the Ensure parameter should not be mandatory. But the Display Name instead.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.schema.mof [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+
+[ClassVersion("1.0.0.0"), FriendlyName("AADEntitlementManagementAccessPackageAssignmentPolicy")]
+class MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy : OMI_BaseResource
+{
+    [Write, Description("Id of the access package assignment policy.")] String Id;
+    [Write, Description("Identifier of the access package.")] String AccessPackageId;
+    [Write, Description("Who must review, and how often, the assignments to the access package from this policy. This property is null if reviews are not required."), EmbeddedInstance("MSFT_MicrosoftGraphassignmentreviewsettings")] String AccessReviewSettings;
+    [Write, Description("Indicates whether a user can extend the access package assignment duration after approval.")] Boolean CanExtend;
+    [Write, Description("The description of the policy.")] String Description;
+    [Write, Description("The display name of the policy.")] String DisplayName;
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy/MSFT_AADEntitlementManagementAccessPackageAssignmentPolicy.schema.mof:127]: At least one property should be marked as Key. I think the DisplayName property is most appropriate here.

**Review by ykuijs — CHANGES_REQUESTED** [review]
Few last comments

**Comment by William-Francillette** [comment]
Should be good to go now

**Review by ykuijs — APPROVED** [review]
LGTM

**Pull request merged**